parallel-cluster-post-install.sh

```diff
@@ -68,87 +88,271 @@ sed -i "s|@CUDA_INSTALL_PATH@|${CUDA_INSTALL_PATH}|g" /etc/profile.d/cuda.sh
+## Available variables:
+## {{ run_cmd }}            - the complete command string to run the job
+## {{ num_cpu }}            - the number of CPUs needed
+## {{ num_gpu }}            - the number of GPUs needed.
+##                            Note: the code will use this many GPUs starting from dev id 0
+##                                  the cluster scheduler or this script have the responsibility
+##                                  of setting CUDA_VISIBLE_DEVICES so that the job code ends up
+##                                  using the correct cluster-allocated GPUs.
+## {{ ram_gb }}             - the amount of RAM needed in GB
+## {{ job_dir_abs }}        - absolute path to the job directory
+## {{ project_dir_abs }}    - absolute path to the project dir
+## {{ job_log_path_abs }}   - absolute path to the log file for the job
+## {{ worker_bin_path }}    - absolute path to the cryosparc worker command
+## {{ run_args }}           - arguments to be passed to cryosparcw run
+## {{ project_uid }}        - uid of the project
+## {{ job_uid }}            - uid of the job
+## {{ job_creator }}        - name of the user that created the job (may contain spaces)
+## {{ cryosparc_username }} - cryosparc username of the user that created the job (usually an email)
+## {{ job_type }}           - CryoSPARC job type
+##
+## What follows is a simple SLURM script:
 
-cat > "${CRYOSPARC_INSTALL_PATH}"/cluster_script.sh << 'EOF'
-#!/bin/bash
-#SBATCH --job-name=cryosparc_{{ project_uid }}_{{ job_uid }}
+#SBATCH --job-name cryosparc_{{ project_uid }}_{{ job_uid }}
+#SBATCH -n {{ num_cpu }}
+${SBATCH_EXTRA}
+#SBATCH --partition=${PARTITION}
+#SBATCH --mem={{ (ram_gb|float * ram_gb_multiplier|float)|int }}G
 #SBATCH --output={{ job_log_path_abs }}
 #SBATCH --error={{ job_log_path_abs }}
-#SBATCH --nodes=1
-#SBATCH --ntasks-per-node=1
-#SBATCH --cpus-per-task={{ num_cpu }}
-#SBATCH --gres=gpu:{{ num_gpu }}
-#SBATCH --partition=@PARTITION@
+
 {{ run_cmd }}
 EOF
 
-sed -i "s|@PARTITION@|${PARTITION}|g" "${CRYOSPARC_INSTALL_PATH}"/cluster_script.sh
+    #sed -i "s|@PARTITION@|${PARTITION}|g" "${CRYOSPARC_INSTALL_PATH}"/cluster_script.sh
+    chown -R ${OSUSER}:${OSGROUP} ${CRYOSPARC_INSTALL_PATH}/${PARTITION}
+
+    # Connect CryoSPARC worker nodes to cluster
+    /bin/su -c "cd ${CRYOSPARC_INSTALL_PATH}/${PARTITION} && ${CRYOSPARC_INSTALL_PATH}/cryosparc_master/bin/cryosparcm cluster connect" - ${OSUSER}
+
+    # Individually apply custom_vars
+    CLICMD=$(cat << EOT
+set_scheduler_target_property(hostname="${PARTITION}",key="custom_vars",value={"ram_gb_multiplier": "${PARTITION_RAM_GB_MULTIPLIER}"})
+EOT
+)
+    /bin/su -c "cd ${CRYOSPARC_INSTALL_PATH}/${PARTITION} && ${CRYOSPARC_INSTALL_PATH}/cryosparc_master/bin/cryosparcm cli '$CLICMD' " - ${OSUSER}
+    echo "Done connecting $PARTITION"
+  else
+    echo "Partition already connected to CryoSPARC: ${PARTITION}"
+  fi
+
+done
+set -e
+
+# VALIDATE CRYOSPARC
```

Comment:
Why is this commented out?

Reply:
This validates the CryoSPARC lane configuration, which can be done after cluster creation. Doing it at cluster creation increases probability of a head node wait condition timeout with ParallelCluster. So we commented it out.

Reply:
Let's add a comment to this effect or pull it out into the GitHub README rather than just commenting it out. :)

Reply:
Done.